keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,349 @@
+
+I would like to be able to get this information by inspecting the local queue
+status.
+
+<!--
+#### Story 2
+
+As a user of Kueue with ClusterQueue visibility I would like to know
+the position of my workload in the cluster queue, so that I can estimate the
+ETA of my workload.
+
+I would like to be able to get this information by inspecting the cluster queue
+status.
+-->
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Time and memory consuming computation of top pending workloads
+
+Currently, we organize the cluster queue as a heap, thus only the top of the
+heap is readily available. If we want to get to know the top N pending workloads
+we may need to copy the heap, sort and select N, which is time and memory
+consuming.
+
+In order to mitigate this risk we may need to migrate from using a heap to
+red-black trees.
+
+#### Too large local queue object
+
+As the number of pending workloads is arbitrarily large there is a risk that the
+status information about the workloads may exceed the etcd limit of 1.5Mi on
+object size. In order to allow feature extensions of the structure we
+should assume not more that 500Ki is used.
+
+In order to mitigate this risk we put a constraint on the number of exposed
+pending workloads. We limit the number to 1000.
+
+#### Increased number of localqueue updates
+
+As we put the global cluster queue position into the list of top pending workloads
+changes to one local queue may trigger updates in another local queue connected
+to the same cluster queue, due to status changes in the structure.
+
+In order to mitigate this risk we need to consider:
+1. batching updates to local queues, by a batch period, similarly as Job status
+   updates are batched in Kubernetes.
+2. allow to control the number of N workloads which expose their position in the
+   cluster queue. Setting low value for the limit would reduce the change of
+   changes in one local queue to trigger an update in another local queue.
+
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Local Queue API
+
+```golang
+// PendingWorkload contains the information identifying a pending workload in
+// the local queue.
+type PendingWorkload struct {
+	// Name indicates the name of the pending workload.
+	Name string
+
+	// Position indicates the position of the workload among all pending
+	// workloads in the cluster queue.
+	Position *int32
```

Comment:
we should call this `approximatePosition`, for two reasons:
- by the time the position is posted, it's probably out of date.
- it gives us the chance to use coarse numbers: `1, 2, 3, ..., 10, 20, 30, ..., 100, 200, 300, ..., 1000`. This might potentially reduce the need for updates.

Reply:
Why do we actually need this? Without the access to the name of the other workload, the position is not particularly useful. The user will not know whether there is 1 huge job or 100 tiny before his task in the queue. At the same time, Kueue will be using a lot of APIServer write quota to maintain this information.

Reply:
> we should call this `approximatePosition`, for two reasons:
> 
> * by the time the position is posted, it's probably out of date.

If we batch updates and have 5s it does not seem relevant enough to name it `approximate`. Similarly, we call `failed` counter in Job controller despite some potential delay.

> * it gives us the chance to use coarse numbers: `1, 2, 3, ..., 10, 20, 30, ..., 100, 200, 300, ..., 1000`. This might potentially reduce the need for updates.

I think we want coarse positions, we should probably mention a range rather than a specific positions.

Reply:
Still, what is the point of this number? With large amount of LocalQueue pointing to the same ClusterQueue, the bigger part of the write quota will be used for these updates, even with 5 sec  batching.

Reply:
> Still, what is the point of this number? 

The point is to give the user some feedback about its position in the cluster queue (which the user does not see). This is analogous to the hot-line bots saying "your position is 5". As a caller you also don't know how long it would take for the other people calling to finish, however, you can get some estimate assuming 10min per caller. Similarly, in the context of Kueue you can get a rough estimate on how long it takes to process 5 items based on the historical velocity of the processing. 

I think, ultimately along with the position we would like to have ETA. Getting ETA based on time execution per jobs might be involving (to collect data based on historical data per job). Still, we could have a rough ETA based on overall cluster queue velocity.

Reply:
Honestly, I'm still not convinced (the benefits vs write cost are not clear), but maybe i'm wrong. What about adding it as a possible phase 2 feature?

Reply:
This is a good point, returning positions adds a complication to the implementation to throttle the updates. I think we should indeed start the implementation without positions for localqueues + cluster queues. 

Certainly we should split the implementation into small commits / PRs, when it makes sense. As for the KEP, yes I'm not sure about the phase process. I also could split the KEP into two also (cluster queue KEP, and local queue KEP). 

However, I think returning local queue order without even positioning is pointless. Returning position gives at least some clue, if you can estimate the queue velocity.

Reply:
> Honestly, I'm still not convinced (the benefits vs write cost are not clear), but maybe i'm wrong. What about adding it as a possible phase 2 feature?

I have split the implementation into phases now. PTAL.